src/reactive-membrane.ts

```diff
@@ -76,31 +90,150 @@ const defaultValueMutated: ReactiveMembraneMutationCallback = (obj: any, key: Pr
-            };
+            descriptor.set = reverseSetterMap.get(set) || set;
         }
     }
     return descriptor;
 }
 
+export function copyDescriptorIntoShadowTarget(
+    handler: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget,
+    key: PropertyKey,
+) {
+    const { originalTarget } = handler;
+    // Note: a property might get defined multiple times in the shadowTarget
+    //       but it will always be compatible with the previous descriptor
+    //       to preserve the object invariants, which makes these lines safe.
+    const normalizedDescriptor = getOwnPropertyDescriptor(originalTarget, key);
+    if (!isUndefined(normalizedDescriptor)) {
+        const blueDesc = wrapDescriptor(handler, normalizedDescriptor);
+        ObjectDefineProperty(shadowTarget, key, blueDesc);
+    }
+}
+
+export function lockShadowTarget(
+    handler: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget
+): void {
+    const { originalTarget } = handler;
+    const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
+    targetKeys.forEach((key: PropertyKey) => {
+        copyDescriptorIntoShadowTarget(handler, shadowTarget, key);
+    });
+
+    preventExtensions(shadowTarget);
+}
+
+export function isExtensibleMembraneTrap(
+    this: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget
+): boolean {
+    const { originalTarget } = this;
+    // optimization to avoid attempting to lock down the shadowTarget multiple times
+    if (!isExtensible(shadowTarget)) {
+        return false; // was already locked down
+    }
+    if (!isExtensible(originalTarget)) {
+        lockShadowTarget(this, shadowTarget);
+        return false;
+    }
+    return true;
+}
+
+export function getOwnPropertyDescriptorMembraneTrap(
+    this: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget,
+    key: PropertyKey
+): PropertyDescriptor | undefined {
+    const { originalTarget, membrane: { valueObserved } } = this;
+
+    // keys looked up via hasOwnProperty need to be reactive
+    valueObserved(originalTarget, key);
+
+    const desc = getOwnPropertyDescriptor(originalTarget, key);
+    if (isUndefined(desc)) {
+        return desc;
+    }
+
+    if (desc.configurable === false) {
+        // updating the descriptor to non-configurable on the shadow
+        copyDescriptorIntoShadowTarget(this, shadowTarget, key);
+    }
+    // Note: by accessing the descriptor, the key is marked as observed
+    // but access to the value, setter or getter (if available) cannot observe
+    // mutations, just like regular methods, in which case we just do nothing.
+    return wrapDescriptor(this, desc);
+}
+
+export function preventExtensionsMembraneTrap(
+    this: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget
+): boolean {
+    const { originalTarget } = this;
+    if (isExtensible(shadowTarget)) {
```

Comment:
improve it to support proxy as targets, something that wasn't really work before.